src/Workspaces/Core/Portable/Options/LegacyWorkspaceOptionService.cs

```diff
@@ -0,0 +1,98 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Options;
+
+[ExportWorkspaceService(typeof(ILegacyWorkspaceOptionService)), Shared]
+internal sealed class LegacyWorkspaceOptionService : ILegacyWorkspaceOptionService
+{
+    public IGlobalOptionService GlobalOptions { get; }
+
+    // access is interlocked
+    private ImmutableArray<Workspace> _registeredWorkspaces;
+
+    /// <summary>
+    /// Stores options that are not defined by Roslyn and do not implement <see cref="IOption2"/>.
+    /// </summary>
+    private ImmutableDictionary<OptionKey, object?> _currentExternallyDefinedOptionValues;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public LegacyWorkspaceOptionService(IGlobalOptionService globalOptionService)
+    {
+        GlobalOptions = globalOptionService;
+        _registeredWorkspaces = ImmutableArray<Workspace>.Empty;
+        _currentExternallyDefinedOptionValues = ImmutableDictionary.Create<OptionKey, object?>();
+    }
+
+    public object? GetOption(OptionKey key)
+    {
+        if (key.Option is IOption2 internallyDefinedOption)
+        {
+            return GlobalOptions.GetOption<object?>(new OptionKey2(internallyDefinedOption, key.Language));
+        }
+
+        if (_currentExternallyDefinedOptionValues.TryGetValue(key, out var value))
+        {
+            return value;
+        }
+
+        return key.Option.DefaultValue;
+    }
+
+    /// <summary>
+    /// Sets values of options that may be stored in <see cref="Solution.Options"/> (public options).
+    /// Clears <see cref="SolutionOptionSet"/> of registered workspaces so that next time
+    /// <see cref="Solution.Options"/> are queried for the options new values are fetched from 
+    /// <see cref="GlobalOptionService"/>.
+    /// </summary>
+    public void SetOptions(
+        ImmutableArray<KeyValuePair<OptionKey2, object?>> internallyDefinedOptions,
+        ImmutableArray<KeyValuePair<OptionKey, object?>> externallyDefinedOptions)
+    {
+        var anyOptionChanged = GlobalOptions.SetGlobalOptions(internallyDefinedOptions);
+
+        foreach (var (optionKey, value) in externallyDefinedOptions)
+        {
+            var existingValue = GetOption(optionKey);
+            if (Equals(value, existingValue))
+            {
+                continue;
+            }
+
+            anyOptionChanged = true;
+            ImmutableInterlocked.GetOrAdd(ref _currentExternallyDefinedOptionValues, optionKey, value);
+        }
+
+        if (anyOptionChanged)
+        {
+            UpdateRegisteredWorkspaces();
+        }
+    }
+
+    private void UpdateRegisteredWorkspaces()
```

Comment:
wouldn't it be necessary that changes to global options trigger this?  It looks like in `SolutionOptionSet` we cache the value of the option retrieved by this type (which could be a global option since it looks like we're asking the global option service).  So if that global option changes the SolutionOptionSet wouldn't see it if the option value had been read and cached once.

Reply:
This is actually problematic. 

Previously we had two code paths - if a global option was changed via `SetOptions` or `RefreshOption` we updated the current solution of each registered workspace. These APIs are called e.g. from `Workspace.TryApplyChanges` and VS tools options UI. The second code path was via `SetGlobalOption`, which did *not* update the registered workspaces. The problem of changing the latter to update the workspace is that pattern like this would break:

```
var solution = workspace.CurrentSolution;
globalOptions.SetGlobalOption(...);
workspace.TryApplyChanges(solution.With...);
```

This is because SetGlobalOption now changes the version of the solution in the workspace and TryApplyChanges no longer works. This is unexpected. Global options should not have impact on workspace in the new world. The fact that they do is unfortunate legacy behavior.

So I'm gonna keep the semantics and only update the solutions on the former code paths. Setting internally defined public global options would then have no effect on their values in Workspace.Options.